### PR TITLE
Laboratory work #3, Viktoriya Aleksina - 23FPL2

### PR DESCRIPTION
**Pull request by AleksinaV**

### CODE REVIEW

File: lab_3_ann_retriever/main.py [hunk 1]
```diff
@@ -49,6 +51,12 @@ def calculate_distance(query_vector: Vector, document_vector: Vector) -> float |
 
     In case of corrupt input arguments, None is returned.
     """
+    dist = 0
+    i = 0
+    while i < len(query_vector):
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:56]: I prefer `for` loop over `zip` of `query_vector` and `document_vector`

File: lab_3_ann_retriever/main.py [hunk 2]
```diff
@@ -104,6 +113,14 @@ def tokenize(self, text: str) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(text, str):
+            return None
+
+        for symbol in text:
+            if not symbol.isalpha():
+                text = text.replace(symbol, ' ')
+        tokenized_text = self._remove_stop_words(text.lower().split())
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:122]: Why do you need this variable?

File: lab_3_ann_retriever/main.py [hunk 3]
```diff
@@ -130,6 +155,15 @@ def _remove_stop_words(self, tokens: list[str]) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if (not tokens or
+                not isinstance(tokens, list) or
+                not all(isinstance(token, str) for token in tokens)):
+            return None
+
+        for token in tokens.copy():
+            if token in self._stop_words or token == 'её':
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:164]: ? `её` is a good word. What's wrong?

File: lab_3_ann_retriever/main.py [hunk 4]
```diff
@@ -149,6 +183,8 @@ def __init__(self, corpus: list[list[str]]) -> None:
         Args:
             corpus (list[list[str]]): Tokenized documents to vectorize
         """
+        self._corpus = corpus
+        self.build()
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:187]: I do not think that you should call `build` there

File: lab_3_ann_retriever/main.py [hunk 5]
```diff
@@ -157,6 +193,13 @@ def build(self) -> bool:
         Returns:
             bool: True if built successfully, False in other case
         """
+        if self._corpus is None or len(self._corpus) == 0:
+            return False
+
+        self._vocabulary = sorted(build_vocabulary(self._corpus))
+        self._idf_values = calculate_idf(self._vocabulary, self._corpus)
+        self._token2ind = {token: self._vocabulary.index(token) for token in self._vocabulary}
+        return True if self._vocabulary and self._idf_values and self._token2ind else False
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:202]: the same as `bool(self._vocabulary and self._idf_values and self._token2ind)`

File: lab_3_ann_retriever/main.py [hunk 6]
```diff
@@ -579,4 +686,4 @@ def __init__(self, vectorizer: Vectorizer, tokenizer: Tokenizer) -> None:
         Args:
             vectorizer (Vectorizer): Vectorizer for documents vectorization
             tokenizer (Tokenizer): Tokenizer for tokenization
-        """
+        """
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:689]: revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_ann_retriever/main.py (hunk 1 above)
Comment (AleksinaV) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py (hunk 2 above)
Comment (AleksinaV) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py (hunk 5 above)
Comment (AleksinaV) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py (hunk 4 above)
Comment (AleksinaV) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (AleksinaV) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py [hunk 7]
```diff
@@ -253,9 +333,20 @@ def index_documents(self, documents: list[str]) -> bool:
...
+            indexed_doc = self._index_document(doc)
+            if not indexed_doc:
+                return False
+            self._documents.append(doc)
+            self._document_vectors.append(indexed_doc)
+        return True
 
     def retrieve_relevant_documents(
-        self, query: str, n_neighbours: int
+            self, query: str, n_neighbours: int
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:349]: revert

File: lab_3_ann_retriever/main.py [hunk 8]
```diff
@@ -304,9 +412,14 @@ def retrieve_vectorized(self, query_vector: Vector) -> str | None:
...
         """
+        if not is_valid_vector(query_vector):
+            return None
+
+        nearest_neighbors = self._calculate_knn(query_vector, self._document_vectors, 1)
+        return self._documents[nearest_neighbors[0][0]]
 
     def _calculate_knn(
-        self, query_vector: Vector, document_vectors: list[Vector], n_neighbours: int
+            self, query_vector: Vector, document_vectors: list[Vector], n_neighbours: int
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:422]: revert

File: lab_3_ann_retriever/main.py [hunk 9]
```diff
@@ -366,11 +500,11 @@ class Node(NodeLike):
     right_node: NodeLike | None
 
     def __init__(
-        self,
-        vector: Vector = (),
-        payload: int = -1,
-        left_node: NodeLike | None = None,
-        right_node: NodeLike | None = None,
+            self,
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:503]: revert

File: lab_3_ann_retriever/settings.json [hunk 10]
```diff
@@ -1,3 +1,3 @@
 {
-    "target_score": 0
+    "target_score": 6
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/settings.json:2]: what about 8?

File: lab_3_ann_retriever/start.py [hunk 11]
```diff
@@ -3,6 +3,7 @@
 """
 
 # pylint:disable=duplicate-code, too-many-locals, too-many-statements, unused-variable
+import lab_3_ann_retriever.main as m
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/start.py:6]: strange import

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_ann_retriever/main.py [hunk 12]
```diff
@@ -5,12 +5,21 @@
...
+from lab_2_retrieval_w_bm25.main import build_vocabulary, calculate_idf, calculate_tf
+from math import sqrt
 from typing import Protocol
 
+
 Vector = tuple[float, ...]
 "Type alias for vector representation of a text."
 
 
+def is_valid_vector(vector: any) -> bool:
```
[the remark this answers is not in the document]
Comment (AleksinaV) [reply lab_3_ann_retriever/main.py:17]: It's sad... Done

File: lab_3_ann_retriever/start.py (hunk 11 above)
Comment (AleksinaV) [reply lab_3_ann_retriever/start.py:6]: Why? I've done this in previous labs too

File: lab_3_ann_retriever/main.py [hunk 13]
```diff
@@ -130,6 +162,15 @@ def _remove_stop_words(self, tokens: list[str]) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if (not tokens or
+                not isinstance(tokens, list) or
+                not all(isinstance(token, str) for token in tokens)):
+            return None
+
+        for token in tokens.copy():
+            if token in self._stop_words:
```
[the remark this answers is not in the document]
Comment (AleksinaV) [reply lab_3_ann_retriever/main.py:171]: done

File: lab_3_ann_retriever/main.py [hunk 14]
```diff
@@ -440,6 +626,12 @@ def query(self, vector: Vector, k: int = 1) -> list[tuple[float, int]] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if (not is_valid_vector(vector) or
+                len(vector) == 0 or not isinstance(k, int) or
+                k <= 0):
+            return None
+        nearest_neighbors = self._find_closest(vector, k)
```
[the remark this answers is not in the document]
Comment (AleksinaV) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py [hunk 15]
```diff
@@ -49,6 +58,13 @@ def calculate_distance(query_vector: Vector, document_vector: Vector) -> float |
 
     In case of corrupt input arguments, None is returned.
     """
+    if not is_valid_vector(query_vector) or not is_valid_vector(document_vector):
+        return None
+
+    dist = 0
+    for que_vec, doc_vec in zip(query_vector, document_vector):
```
[the remark this answers is not in the document]
Comment (AleksinaV) [reply lab_3_ann_retriever/main.py:65]: Sorry, but i don't understand how to do it

File: lab_3_ann_retriever/main.py (hunk 8 above)
Comment (AleksinaV) [reply lab_3_ann_retriever/main.py:428]: done

**Pull request closed without merging**